crates/extricrate/src/lib.rs

```diff
@@ -44,13 +61,86 @@ pub mod dependencies {
 
     pub type UseStatements = Vec<UseStatement>;
 
+    #[derive(Debug, Hash, PartialEq, Eq)]
     pub struct File(String);
 
     pub type UseStatementMap = HashMap<File, UseStatements>;
 
+    struct UseVisitor;
+
+    impl<'ast> Visit<'ast> for UseVisitor {
+        fn visit_item_use(&mut self, node: &'ast ItemUse) {
+            let tokens = node.to_token_stream();
+            dbg!(flatten_use_tree("", &node.tree));
+            visit::visit_item_use(self, node);
+        }
+    }
+
+    fn flatten_use_tree(prefix: &str, tree: &UseTree) -> Vec<UseStatementType> {
+        match tree {
+            UseTree::Path(UsePath { ident, tree, .. }) => {
+                let new_prefix = if prefix.is_empty() {
+                    ident.to_string()
+                } else {
+                    format!("{}::{}", prefix, ident)
+                };
+                flatten_use_tree(&new_prefix, tree)
+            }
+
+            UseTree::Name(UseName { ident, .. }) => {
+                let full = if prefix.is_empty() {
+                    ident.to_string()
+                } else {
+                    format!("{}::{}", prefix, ident)
+                };
+                vec![UseStatementType::Simple(full)]
+            }
+
+            UseTree::Rename(UseRename { ident, rename, .. }) => {
+                let full = if prefix.is_empty() {
+                    ident.to_string()
+                } else {
+                    format!("{}::{}", prefix, ident)
+                };
+                vec![UseStatementType::Alias(full, rename.to_string())]
+            }
+
+            UseTree::Glob(UseGlob { .. }) => {
+                vec![UseStatementType::WildCard]
+            }
+
+            UseTree::Group(UseGroup { items, .. }) => items
+                .iter()
+                .flat_map(|subtree| flatten_use_tree(prefix, subtree))
+                .collect(),
+        }
+    }
+
+    #[derive(Debug, Error)]
+    pub enum ListUseStatementError {
+        #[error("file not found")]
+        FileNotFound,
+        #[error("file not parsable")]
+        FileNotParsable,
+        #[error("file not readable")]
+        FileNotReadable,
+    }
+
     /// List all the `use` statements in the crate, by file/module.
-    pub fn list_use_statements(crate_root: &std::path::Path) -> UseStatementMap {
-        todo!()
+    pub fn list_use_statements(
+        crate_root: &std::path::Path,
+    ) -> Result<UseStatementMap, ListUseStatementError> {
+        if !crate_root.exists() {
+            return Err(ListUseStatementError::FileNotFound);
+        }
+        let content =
+            read_to_string(crate_root).map_err(|_| ListUseStatementError::FileNotReadable)?;
```

Comment:
I feel like crate_root should be a directory, where you find the Cargo.toml, not a source file. That complicates this function a bit, so maybe what you wrote could be a sub-function that gets called for each file, and we can test it separately?